src/app/clusters/actions-server/actions-server.cpp

```diff
@@ -0,0 +1,330 @@
+/*
+ *
+ *    Copyright (c) 2024 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include "actions-server.h"
+#include <app-common/zap-generated/attributes/Accessors.h>
+#include <app-common/zap-generated/cluster-objects.h>
+#include <app-common/zap-generated/ids/Attributes.h>
+#include <app/AttributeAccessInterfaceRegistry.h>
+#include <app/CommandHandlerInterfaceRegistry.h>
+#include <app/util/attribute-storage.h>
+#include <lib/support/CodeUtils.h>
+#include <lib/support/logging/CHIPLogging.h>
+
+using namespace chip;
+using namespace chip::app;
+using namespace chip::app::Clusters;
+using namespace chip::app::Clusters::Actions;
+using namespace chip::app::Clusters::Actions::Attributes;
+
+Instance Instance::instance;
+Instance * Instance::GetInstance()
+{
+    return &instance;
+}
+
+CHIP_ERROR Instance::ReadActionListAttribute(const AttributeValueEncoder::ListEncodeHelper & encoder)
+{
+    if (GetInstance()->mDelegate == nullptr)
+    {
+        ChipLogError(Zcl, "Actions delegate is null!!!");
+        return CHIP_ERROR_INCORRECT_STATE;
+    }
+    for (uint16_t i = 0; true; i++)
+    {
+        Actions::Structs::ActionStruct::Type action;
+
+        CHIP_ERROR err = GetInstance()->mDelegate->ReadActionAtIndex(i, action);
+        if (err == CHIP_ERROR_PROVIDER_LIST_EXHAUSTED)
```

Comment:
Where does the API document that you return CHIP_ERROR_PROVIDER_LIST_EXHAUSTED on end-of-list?  It doesn't, and should.

Reply:
Added documentation.